### DISCUSSION TITLE
Metrics: batch_duration vs process_duration

### DISCUSSION THREAD

**Discussion by secsilm** [mosecorg/mosec#565] [Q&A] [answered]
The document briefly explains the differences between the two

> `mosec_service_batch_duration_second_bucket` shows the duration of dynamic batching for each connection in each stage (starts from receiving the first task).

> `mosec_service_process_duration_second_bucket` shows the duration of processing for each connection in each stage (including the IPC time but excluding the mosec_service_batch_duration_second_bucket).

But I'm still not very clear about the specific stages they represent. Assuming we only have one stage, does `batch_duration` refer to the time from receiving the first request to returning all requests? Then what time does `process_duration` include? What is the relationship between them?

Thank you!

**Comment by kemingy** [comment, marked as answer]
- `batch_duration`: gathering tasks for one batch (count start from receiving the 1st task)
- `process_duration`: from sending the batch to receiving the results of the batch (including IPC time)

You can check the [concept](https://mosec.readthedocs.io/en/latest/reference/concept.html).

> Assuming we only have one stage, does batch_duration refer to the time from receiving the first request to returning all requests? Then what time does process_duration include? What is the relationship between them?

```
requests => batched requests => sending to worker => worker process => receiving from worker => send results to clients
```

`batch_duration` + `process_duration` is roughly (could be less if there are too many requests waiting to be processed) the time spent inside the service.

**Reply by secsilm** [reply]
Okay, I seem to understand. The `batch_duration` refers to the time taken to bundle the requests together, while the `process_duration` refers to the time taken to process that batch. The actual inference time of the model is included in the `process_duration`.

Am I understanding it correctly?

**Reply by kemingy** [reply]
Yes, your understanding is correct. `process_duration` includes model inference and IPC time.
